src/plugins/memdump/userhook.cpp

```diff
@@ -122,20 +122,215 @@
+                    "\"UserName\": \"%s\", "
+                    "\"UserId\": %" PRIu64 ", "
+                    "\"PID\": %d, "
+                    "\"PPID\": %d, "
+                    "\"Method\": \"%s\", "
+                    "\"CalledFrom\": 0x%" PRIx64 ", "
+                    "\"Arguments\": [",
+                    UNPACK_TIMEVAL(info->timestamp),
+                    escaped_pname,
+                    USERIDSTR(drakvuf), info->proc_data.userid,
+                    info->proc_data.pid, info->proc_data.ppid,
+                    info->trap->name,
+                    info->regs->rip);
+
+            print_arguments(ret_target->arguments);
+            printf("}");
+            g_free(escaped_pname);
+            break;
+        default:
+        case OUTPUT_DEFAULT:
+            printf("[MEMDUMP-USERHOOK] TIME:" FORMAT_TIMEVAL " VCPU:%" PRIu32 " CR3:0x%" PRIx64 " ProcessName:\"%s\" UserID:%" PRIi64 " Method:\"%s\" CalledFrom:0x%" PRIx64 " Arguments:\"",
+                   UNPACK_TIMEVAL(info->timestamp), info->vcpu, info->regs->cr3, info->proc_data.name,
+                   info->proc_data.userid, info->trap->name, info->regs->rip);
+            print_arguments(ret_target->arguments);
+            printf("\"");
+            break;
+    }
+    printf("\n");
+
+    drakvuf_remove_trap(drakvuf, info->trap, nullptr);
+	delete ret_target;
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
 static event_response_t usermode_hook_cb(drakvuf_t drakvuf, drakvuf_trap_info* info)
 {
     hook_target_entry_t* target = (hook_target_entry_t*)info->trap->data;
 
     if (target->pid != info->proc_data.pid)
         return VMI_EVENT_RESPONSE_NONE;
 
-    PRINT_DEBUG("[MEMDUMP-USER] Usermode callback called %d!%s\n", info->proc_data.pid, info->trap->name);
     dump_from_stack(drakvuf, info, target->plugin);
+    bool is_syswow = is_wow64(drakvuf, info);
+
+    access_context_t ctx =
+            {
+                    .translate_mechanism = VMI_TM_PROCESS_DTB,
+                    .dtb = info->regs->cr3,
+                    .addr = info->regs->rsp
+            };
+
+    bool success = false;
+    addr_t ret_addr = 0;
+
+    auto vmi = drakvuf_lock_and_get_vmi(drakvuf);
+    if (is_syswow)
+    {
+        uint32_t ret_addr_tmp;
+
+        if (vmi_read_32(vmi, &ctx, &ret_addr_tmp) == VMI_SUCCESS)
+        {
+            success = true;
+            ret_addr = ret_addr_tmp;
+        }
+    }
+    else
+    {
+        success = vmi_read_64(vmi, &ctx, &ret_addr) == VMI_SUCCESS;
+    }
+    drakvuf_release_vmi(drakvuf);
+
+	return_hook_target_entry_t* ret_target = new return_hook_target_entry_t();
+    drakvuf_trap_t* trap = new drakvuf_trap_t;
+
+	for (size_t i = 1; i <= target->args_num; i++)
+	{
+		uint64_t argument = get_function_argument(drakvuf, info, i);
+		ret_target->arguments.push_back(argument);
+	}
+	ret_target->plugin = target->plugin;
+
+	addr_t paddr;
+	vmi = drakvuf_lock_and_get_vmi(drakvuf);
+    vmi_v2pcache_flush(vmi, info->regs->cr3);
```

Comment:
You might as well flush the cache earlier if you think it might be stale. You already have read some values through this cr3 above.